### PR TITLE
Add discovery server and fix controller JSON mapping

### PR DESCRIPTION
**Pull request by jirassimok**
This should be merged after pull request #25.

These two commits don't really belong in the same commit. We need to get a lot of more-or-less unrelated changes through, and pull request #30 depends on both of these commits.

### CODE REVIEW

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java [hunk 1]
```diff
@@ -35,16 +40,30 @@ protected void setup(Binder binder)
         binder.disableCircularProxies();
 
         binder.bind(NodeSet.class).in(Scopes.SINGLETON);
+        newExporter(binder).export(NodeSet.class).withGeneratedName();
```
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: What is the purpose of this new exporter, this is more like a question.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java [hunk 2]
```diff
@@ -35,16 +40,30 @@ protected void setup(Binder binder)
...
     @Provides
     public JerseyClient jerseyClientProvider()
     {
         return JerseyClientBuilder.createClient();
     }
+
+    @Provides
+    @Singleton
+    @ForController
+    public ObjectMapper obectMapper()
```
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: Misspelling. And it it not used anywhere.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java [hunk 3]
```diff
@@ -13,40 +13,46 @@
...
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
 public class NodeSet
 {
-    private Set<Agent> nodeSet = ConcurrentHashMap.newKeySet();
+    private static final Logger LOG = Logger.get(NodeSet.class);
```
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: We should name the logger to be `LOG` or `LOGGER` always

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java [hunk 4]
```diff
@@ -74,7 +82,40 @@ public void removeAgent(UUID id)
...
+                        properties.get("http"), id.toString());
+                /* TODO: Map this exception to a specific response from the server with...
+                 * ExceptionMapper, or change it to a WebApplicationException containing
+                 * a Response (maybe via a utility class). If using ExceptionMapper, be
+                 * careful not to catch other DiscoveryExceptions (i.e. change this one).
+                 */
+                throw new DiscoveryException("", e);
+            }
+            Agent agent = new Agent(uri, isCoordinator, isWorker, id);
+            setBuilder.add(agent);
```
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: Could inline these

File: presto-manager-controller/pom.xml [hunk 5]
```diff
@@ -24,6 +24,12 @@
             <artifactId>presto-manager-common</artifactId>
             <version>${project.version}</version>
         </dependency>
+
+        <dependency>
+            <groupId>io.airlift.discovery</groupId>
+            <artifactId>discovery-server</artifactId>
+            <version>1.28</version>
+        </dependency>
```
Comment (jirassimok) [inline presto-manager-controller/pom.xml]: We don't need it in the Agent, so I put it here.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java (hunk 1 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: That's a good question.

Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: That's a good question.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServer.java [hunk 6]
```diff
@@ -33,16 +36,21 @@ public static void main(String[] args)
...
                 new DiscoveryModule(),
+                new DiscoveryServerModule(),
+                new HttpEventModule(),
                 new HttpServerModule(),
                 new JsonModule(),
-                new JaxrsModule(true), // requireExplicitBindings = true
-                new HttpEventModule(),
+                new JaxrsModule(true), // requireExplicitBindings = true'
+                new JmxModule(),
+                new MBeanModule(),
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServer.java]: The order of the modules is sort-of arbitrary. The Node and Discovery modules are first, as in the airlift samples and Presto, then the build-in modules in alphabetical order, then the single custom module.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java (hunk 2 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: That should be fixed at some point.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java (hunk 3 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: I think it's fine either way.

(I prefer `LOG` because it's shorter, and then you can read "LOG.error" as "log error".)

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java (hunk 4 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: I could. I'll add that to one of my upcoming PRs that modifies this file.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER, METHOD})
+@BindingAnnotation
+public @interface ForController
```
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java]: Can use a better name instead of `ForController`

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java (hunk 3 above)
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: Then it should be changed everywhere. So it's better to use `LOGGER` for uniformity.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java (hunk 7 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java]: I tried to think of one, but couldn't, and "For X" seems to be a common naming convention for binding annotations in Airlift.

This particular annotation is used to for injecting something exclusively into the Presto Manager code, when the class being injected already has a binding from airlift.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java (hunk 3 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: All right, but I'd prefer to put this change off for a separate commit, where we change all of the `Logger`s to be named `LOGGER`.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java (hunk 7 above)
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java]: Are we planning to add more bindings to this class?
Currently I believe you are planning to bind JSON mapper.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java [hunk 8]
```diff
@@ -13,40 +13,46 @@
...
+    private static final Logger LOG = Logger.get(NodeSet.class);
 
-    @Inject
-    public NodeSet() {}
-
-    public void addAgent(URI uri, boolean isCoordinator, boolean isWorker, UUID id)
-    {
-        nodeSet.add(new Agent(uri, isCoordinator, isWorker, id));
-    }
+    private Set<Agent> nodeSet = ImmutableSet.of();
```
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: Is there a check to see whether `nodeSet` is empty?
Or `nodeSet` being empty is not a problem.

Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: There are potential future situations in which we might operate on `nodeSet` without calling `refresh` first. This protects against those situations at virtually no cost.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java (hunk 7 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ForController.java]: We might in the future, but we do not have any specific plans.

(Though we do have plans to do something similar in the Agent.)

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java [hunk 9]
```diff
@@ -40,11 +44,24 @@ protected void setup(Binder binder)
 
         jaxrsBinder(binder).bind(ControllerConfigAPI.class);
         jaxrsBinder(binder).bind(ControllerConnectorAPI.class);
+
+        discoveryBinder(binder).bindSelector("presto-manager");
     }
 
     @Provides
     public JerseyClient jerseyClientProvider()
```
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: Add `@Singleton`

Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/ControllerServerModule.java]: Will do.

File: presto-manager-controller/pom.xml [hunk 10]
```diff
@@ -24,6 +24,12 @@
             <artifactId>presto-manager-common</artifactId>
             <version>${project.version}</version>
         </dependency>
+
+        <dependency>
+            <groupId>io.airlift.discovery</groupId>
+            <artifactId>discovery-server</artifactId>
+            <version>1.28</version>
```
Comment (jirassimok) [inline presto-manager-controller/pom.xml]: The discovery server comes from a maven group, so no.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java [hunk 11]
```diff
@@ -74,7 +82,40 @@ public void removeAgent(UUID id)
...
+    {
+        List<ServiceDescriptor> services = serviceSelector.selectAllServices();
+
+        ImmutableSet.Builder<Agent> setBuilder = ImmutableSet.builder();
+
+        for (ServiceDescriptor service : services) {
+            Map<String, String> properties = service.getProperties();
+            // TODO: Make agents start without an ID, and provide one on first discovery
+            UUID id = service.getId();
+            boolean isCoordinator = Boolean.parseBoolean(properties.get("presto-coordinator"));
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: What do I not need within the loop? Every value assigned in the loop is used in the loop and replaced in the next iteration.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java [hunk 12]
```diff
@@ -65,32 +69,39 @@ public Response forwardRequest(String scope, ApiRequester apiRequester, Collecti
...
+                    .entity("Number of coordinator is not 1").build();
         }
 
         List<Response> responseList = new ArrayList<>();
         for (URI uri : uriCollection) {
             responseList.add(apiRequester.send(uri));
         }
+        List<ResponseWrapper> responses = responseList.stream()
+                .map(ResponseWrapper::wrapResponse)
+                .collect(Collectors.toList());
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java]: There is no existing `Collector` for immutable lists. The easiest solution would be to make an immutable copy of the list after collecting into it, but that seems like a bit much for a list that will be used exactly once. (I could also collect it into an `ImmutableList.Builder`, but that looks awful.)

More importantly, my next pull request rewrites the `for`-loop and stream above entirely. If this section needs immutability, I'd rather add it then than now.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java [hunk 13]
```diff
@@ -40,18 +38,24 @@
...
         this.nodeSet = nodeSet;
+        this.objectMapper = objectMapper;
     }
 
-    public Response forwardRequest(String scope, ApiRequester apiRequester, Collection<UUID> nodeId)
+    public Response forwardRequest(
+            String scope, ApiRequester apiRequester, Collection<UUID> nodeId)
     {
-        if (((scope != null) && (!nodeId.isEmpty())) || (scope == null && nodeId.isEmpty())) {
+        if (((scope != null) && (!nodeId.isEmpty()))
```
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java]: Oops, sorry about that.

**Comment by jirassimok** [comment]
@maciejgrzybek I responded to your comments.

As for the squashing of the fixup commits, the only reason that commit is separate is because there are branches that aren't mine off of the earlier commits in this pull requests. If I change those, we're liable to end up with some unnecessarily messy merge conflicts.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java (hunk 11 above)
Comment (maciejgrzybek) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java]: Sorry, I misread it. Please disregard my previous comment.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java (hunk 12 above)
Comment (maciejgrzybek) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java]: How about `ImmutableList.toImmutableList()` from Guava? I thought you have Guava available in this module.

Anyway, given your next PR is open, we can skip that for now.

Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/RequestDispatcher.java]: Well, that's what I get for looking at StackOverflow instead of the Guavadocs.\

I'll make the rewrite immutable.
